File: 백트래킹/지우/BOJ_1174_줄어드는 수2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int arr[] = {9,8,7,6,5,4,3,2,1,0};
vector<long long> dp;

void dfs(int start, long long sum) {
    if(start != 0) {
        dp.push_back(sum);
    }
    if(start == 10) return;
    
    for(int i=start; i<10; i++) {
        dfs(i+1, sum*10 + arr[i]);
    }
}

int main() {
    int N; cin >> N;

    dfs(0, 0);
    sort(dp.begin(), dp.end());
    int maxN = dp.size();

    if(maxN >= N) cout << dp[N-1];
    else cout << -1;
    return 0;
}

File: 백트래킹/나영/BOJ_1174_줄어드는 수2.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static Scanner sc = new Scanner(System.in);
    static int n;
    static int [] arr = new int [10];
    static List<Long> list = new ArrayList<>();
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) {

        for (int i = 8; i >= 0; i--) {
            arr[i] = 9 - i;
        }
        
        dfs(0, 0);

        list.sort(null);
        
        n = sc.nextInt();
        
        if (n <= list.size()) System.out.println(list.get(n-1));
        else System.out.println(-1);
    }

    static void dfs(long sum, int idx) {
        if (idx != 0) list.add(sum);

        for (int i = idx; i < 10; i++) {
            dfs(sum*10 + arr[i], i+1);
        }
        
    }
}